patroni/config_generator.py

```diff
@@ -0,0 +1,290 @@
+
+    su_params: Dict[str, str] = {}
+    parsed_dsn = {}
+
+    if dsn:
+        parsed_dsn = parse_dsn(dsn)
+        if not parsed_dsn:
+            sys.exit('Failed to parse DSN string')
+
+    # gather auth parameters for the superuser config
+    for conn_param, env_var in AUTH_ALLOWED_PARAMETERS_MAPPING.items():
+        val = parsed_dsn.get(conn_param, os.getenv(env_var))
+        if val:
+            su_params[conn_param] = val
+    patroni_env_su_username = ((config.get('authentication') or {}).get('superuser') or {}).get('username')
+    patroni_env_su_pwd = ((config.get('authentication') or {}).get('superuser') or {}).get('password')
+    # because we use "username" in the config for some reason
+    su_params['username'] = su_params.pop('user', patroni_env_su_username) or getuser()
+    su_params['password'] = su_params.get('password', patroni_env_su_pwd) or getpass('Please enter the user password:')
+
+    from . import psycopg
+    try:
+        conn = psycopg.connect(dsn=dsn, password=su_params['password'])
+    except psycopg.Error as e:
+        sys.exit(f'Failed to establish PostgreSQL connection: {e}')
+
+    with conn.cursor() as cur:
+        cur.execute("SELECT 1 FROM pg_roles WHERE rolname=%s AND rolsuper", (su_params['username'],))
+        if cur.rowcount < 1:
+            sys.exit('The provided user does not have superuser privilege')
+
+        required_params = ['hba_file', 'ident_file', 'config_file',
+                           'data_directory'] + list(ConfigHandler.CMDLINE_OPTIONS.keys())
+        cur.execute("SELECT name, current_setting(name) FROM pg_settings "
+                    "WHERE context <> 'internal' "
+                    "AND source IN ('configuration file', 'command line', 'environment variable') "
+                    "AND category <> 'Write-Ahead Log / Recovery Target' "
+                    "AND setting <> '(disabled)' "
+                    "OR name = ANY(%s)", (required_params,))
+
+        helper_dict = dict.fromkeys(['port', 'listen_addresses'])
+        # adjust values
+        for p, v in cur.fetchall():
+            if p == 'data_directory':
+                config['postgresql']['data_dir'] = v
+            elif p == 'cluster_name' and v:
+                config['scope'] = v
+            elif p in ('archive_command', 'restore_command', 'archive_cleanup_command',
+                       'recovery_end_command', 'ssl_passphrase_command',
+                       'hba_file', 'ident_file', 'config_file'):
+                # write commands to the local config due to security implications
+                # write hba/ident/config_file to local config to ensure they are not removed later
+                config['postgresql'].setdefault('parameters', {})
+                config['postgresql']['parameters'][p] = v
+            elif p in helper_dict:
+                helper_dict[p] = v
+            else:
+                config['bootstrap']['dcs']['postgresql']['parameters'][p] = v
```

Comment:
I was wondering if shouldn't everything other than what is listed in [PostgreSQL parameters controlled by Patroni](https://patroni.readthedocs.io/en/latest/patroni_configuration.html#postgresql-parameters-controlled-by-patroni) go to local configuration?

I mean, I see "dynamic configuration" as the configuration that applies to the whole cluster, and local configuration the configuration that applies only to the local node.

With that in mind, I'd say to write the settings that Patroni thinks we should enforce globally into dynamic configuration, and everything else go to local configuration by default. Then it's up to the user to check what he wants to be applied globally or not.

What do you think of that?

Reply:
What is your use case of having different configuration on different nodes?

Reply:
Maybe we should just keep it as it is. It was just some random thought based on the conception that I got about how these settings apply to nodes.

I already saw people with different settings, like `work_mem`, because queries that were run in the standbys had different aspects than the ones being run in the primary. But in fact, as they are part of a replication scenario, and standbys could take the primary role at any moment, I would prefer the standbys to be similar in terms of hardware and configuration.

In practice it would not change the result -- unless the existing cluster has a divergence in configuration among the nodes. Also, I guess it could make it harder to manage the cluster after that if one attempts to change configuration through `patronictl edti-config`, as the local configuration would take precedence.

Just forget about it 😄